### PR TITLE
MODGOBI-72- Check for existing order before creating

### PR DESCRIPTION
**Pull request by KVupp**
## Purpose
https://issues.folio.org/browse/MODGOBI-72 - Prevent duplicate Orders

## Approach
1) Check for existing Order, and return PO Line Number if its existing
2) Retry to Open Order if its Pending
3) Create Order if any of the above steps fail, or order doesn't exist

## Pre-Merge Checklist
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [x] Code coverage on new code is 80% or greater
  - [x] Duplications on new code is 3% or less
  - [x] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added, or removed?
  - [ ] Are there new interface dependencies?
  - [x] There are no breaking changes in this PR.
  
If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all they appropriate links to blocked/related issues?
- Are the JIRAs under active development?  
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?

Ideally all of the PRs involved in breaking changes would be merged in the same day to avoid breaking the folio-testing environment.  Communication is paramount if that is to be achieved, especially as the number of intermodule and inter-team dependencies increase.  

While it's helpful for reviewers to help identify potential problems, ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

File: src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java [hunk 1]
```diff
@@ -614,6 +621,223 @@ public final void testPostGobiOrdersFallBackLocation(TestContext context) throws
...
+        .extract()
+          .body()
+            .as(GobiResponse.class, ObjectMapperType.JAXB);
+
+    //should fetch existing order which is in Pending state
+    List<JsonObject> getOrder = MockServer.serverRqRs.get(PURCHASEORDER, HttpMethod.GET);
+    assertEquals(1, getOrder.size());
+
+    //should try to Open, the existing Pending Order, which fails
+    List<JsonObject> putOrder = MockServer.serverRqRs.get(COMPOSITE_PURCHASE_ORDER, HttpMethod.PUT);
```
Comment (javalkar-varun) [inline src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java]: As per test description, it should fail to transition Order to Open status so `putOrder` should be `null` but it is not?

File: src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java [hunk 2]
```diff
@@ -614,6 +621,223 @@ public final void testPostGobiOrdersFallBackLocation(TestContext context) throws
...
+    assertEquals(1, putOrder.size());
+
+    //should get the existing Order Line Number
```
Comment (javalkar-varun) [inline src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java]: This is only getting an existing pending order so maybe description should say that.

File: src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java (hunk 1 above)
Comment (KVupp) [reply src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java]: The test is to check if the existing POLine Number is returned even if the PUT fails, so the verification here is just to check if PUT was called ( which is mocked to fail)

File: src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java [hunk 3]
```diff
@@ -614,6 +621,223 @@ public final void testPostGobiOrdersFallBackLocation(TestContext context) throws
...
+
+    //should not try to create a new Order
+    List<JsonObject> postedOrder = MockServer.serverRqRs.get(PURCHASEORDER, HttpMethod.POST);
+    assertNull(postedOrder);
+
+    //return the existing PO Line Number
+    assertNotNull(order.getPoLineNumber());
+
+
+    logger.info("End: Testing for 201 - Return existing Order even if it is pending, and retry to Open suceeds");
```
Comment (javalkar-varun) [inline src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java]: nitpick.. Typo.. `succeeds`

File: src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java [hunk 4]
```diff
@@ -614,6 +621,223 @@ public final void testPostGobiOrdersFallBackLocation(TestContext context) throws
...
+    List<JsonObject> getOrder = MockServer.serverRqRs.get(PURCHASEORDER, HttpMethod.GET);
+    assertEquals(1, getOrder.size());
+
+    //should not retry to Open
+    List<JsonObject> putOrder = MockServer.serverRqRs.get(COMPOSITE_PURCHASE_ORDER, HttpMethod.PUT);
+    assertNull(putOrder);
+
+    //should try to create a new Order
+    List<JsonObject> postedOrder = MockServer.serverRqRs.get(PURCHASEORDER, HttpMethod.POST);
+    assertEquals(1,postedOrder.size());
```
Comment (javalkar-varun) [inline src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java]: space missing between `expected` and `actual`

File: src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java [hunk 5]
```diff
@@ -614,6 +621,223 @@ public final void testPostGobiOrdersFallBackLocation(TestContext context) throws
...
+    assertNotNull(order.getPoLineNumber());
+
+
+    logger.info("End: Testing for 201 - Create new Order , if fetching existing PO Line Number Fails");
```
Comment (javalkar-varun) [inline src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java]: Extra space before comma ` ,`

File: src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java (hunk 3 above)
Comment (KVupp) [reply src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java]: Done

File: src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java (hunk 4 above)
Comment (KVupp) [reply src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java]: Done

File: src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java (hunk 5 above)
Comment (KVupp) [reply src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java]: Done

File: src/test/resources/MockData/purchaseOrders.json [hunk 6]
```diff
@@ -0,0 +1,20 @@
+{
```
Comment (javalkar-varun) [inline src/test/resources/MockData/purchaseOrders.json]: Spacing looks off here. Looks like these are tabs.

File: src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java (hunk 2 above)
Comment (KVupp) [reply src/test/java/org/folio/rest/impl/GOBIIntegrationServiceResourceImplTest.java]: the line 750 has the required comments as it is the place , however added the comment here too

File: src/test/resources/MockData/purchaseOrders.json (hunk 6 above)
Comment (KVupp) [reply src/test/resources/MockData/purchaseOrders.json]: Done

**Pull request merged**